### PR TITLE
Improve QGIS Server WMS GetCapabilities output for root layer <Title> and <Abstract>

### LINKED ISSUE #40736: WMS title and abstract is not displayed for WMS provided by QGIS Server
**Describe the bug**
WMS title is not displayed in QGIS browser for WMS provided by QGIS Server.

**How to Reproduce**

1. Create a new WMS connection using the URL: http://168.119.115.196/cgi-bin/missing_title/qgis_mapserv.fcgi
2. Open the connection in QGIS browser
3. See error --> Title is not displayed:

Using a browser to request the GetCapabilities document, the title is there:
http://168.119.115.196/cgi-bin/missing_title/qgis_mapserv.fcgi?service=WMS&version=1.3.0&request=GetCapabilities

**QGIS and OS versions**

QGIS version | 3.16.2-Hannover | QGIS code revision | f1660f9da5
-- | -- | -- | --
Compiled against Qt | 5.11.3 | Running against Qt | 5.11.3
Compiled against GDAL/OGR | 2.4.0 | Running against GDAL/OGR | 2.4.0
Compiled against GEOS | 3.7.1-CAPI-1.11.1 | Running against GEOS | 3.7.1-CAPI-1.11.1 27a5e771
Compiled against SQLite | 3.27.2 | Running against SQLite | 3.27.2
PostgreSQL Client Version | 11.9 (Debian 11.9-0+deb10u1) | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.4 | QScintilla2 Version | 2.10.4
Compiled against PROJ | 5.2.0 | Running against PROJ | Rel. 5.2.0, September 15th, 2018
OS Version | Debian GNU/Linux 10 (buster)
Active python plugins | ClipMultipleLayers; quick_map_services; scriptrunner3; LAStools; canvas_clipper; savedialog; loadthemall; clipper; kmltools; qgis-maptiler-plugin; MapsPrinter; Discovery; mmqgis; pluginbuilder3; vector_tiles_reader; layerGroupFilter; plugin_reloader; realcentroid; mask; QuickOSM; MetaSearch; db_manager; processing

**Additional context**

[project_and_data_used_for_qgis_server.zip](https://github.com/qgis/QGIS/files/5734481/project_and_data_used_for_qgis_server.zip)

---

For other WMS the title is displayed in QGIS browser, e.g.:
https://www.wms.nrw.de/geobasis/wms_nw_dop

https://www.wms.nrw.de/geobasis/wms_nw_dop?service=WMS&version=1.3.0&request=GetCapabilities

### PR DESCRIPTION
**Pull request by pathmapper**
## Description

This PR improves the QGIS Server WMS GetCapabilities output for root layer `<Title>` and `<Abstract>`.

Fixes #40736

Before, if available, the project title was used for root layer `<Title>` and also `<Abstract>`. If there was no project title available, both elements were missing for the WMS GetCapabilities output. The result were blank fields in QGIS Desktop browser and "add WMS layer" dialogue (#40736). Also it was not possible to set an abstract different than the project title. If I understood it correct, a missing  `<Title>` for the root layer is also a violation of the OGC WMS spec.

Title and abstract from _Service Capabilities_ are preferred over _Project title_, because the assumption is that a user would expect to configure the GetCapabilities output from _Service Capabilities_ (as it's already the case for root layer `<Name>` with _Short name_) and not from the _General_ tab in project properties.
 
This PR implements the following logic:

**root layer `<Title>`**
If _Project → Properties → QGIS Server → Service Capabilities → Title_ is available, it is used for root layer `<Title>`. 
If it's not available, we use the project's title (_Project → Properties → General → Project title_).
Only if project's title is also not set, we use `<Title>untitled</Title>`.

**root layer `<Abstract>`**
If _Project → Properties → QGIS Server → Service Capabilities → Abstract_ is available, it is used for root layer `<Abstract>`. 
If it's not available, `<Abstract>` for the root layer is not included in the GetCapabilities output (according to the WMS spec `<Abstract>` is optional).

### CODE REVIEW

**Comment by Gustry** [comment]
CC @rldhont

**Comment by elpaso** [comment]
Can you move this logic into a method? I think it can be re-used for WFS and OAPIF too. 
There is a QgsServerProjectUtils::owsServiceTitle and abstract functions that look like the idea place where to put this logic.

**Comment by pathmapper** [comment]
> Can you move this logic into a method? 

@elpaso thanks for your feedback, done for _title_ in https://github.com/qgis/QGIS/pull/40758/commits/91d20df94a08d3e86295ddbda953abae07ff8fa1 . For _abstract_ there's currently no fallback used, so nothing to move.

Travis failed unrelated because the docker pull rate limit was reached:

> Service 'postgres' failed to build: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit

File: src/server/qgsserverprojectutils.cpp [hunk 1]
```diff
@@ -37,7 +37,24 @@ bool QgsServerProjectUtils::owsServiceCapabilities( const QgsProject &project )
 
 QString QgsServerProjectUtils::owsServiceTitle( const QgsProject &project )
 {
-  return project.readEntry( QStringLiteral( "WMSServiceTitle" ), QStringLiteral( "/" ) );
+  QString title;
+  QString serviceTitle = project.readEntry( QStringLiteral( "WMSServiceTitle" ), QStringLiteral( "/" ) );
```
Comment (elpaso) [inline src/server/qgsserverprojectutils.cpp:41]: Can't you just assign to `title` here? and check for title empty?

File: src/server/qgsserverprojectutils.cpp [hunk 2]
```diff
@@ -37,7 +37,24 @@ bool QgsServerProjectUtils::owsServiceCapabilities( const QgsProject &project )
...
+  }
+  else
+  {
+    if ( !project.title().isEmpty() )
+    {
+      title = project.title();
+    }
+    else
+    {
+      title = QStringLiteral( "untitled" );
```
Comment (elpaso) [inline src/server/qgsserverprojectutils.cpp:54]:
```suggestion
      title = QObject::tr( "Untitled" );
```

File: src/server/qgsserverprojectutils.cpp (hunk 1 above)
Comment (pathmapper) [reply src/server/qgsserverprojectutils.cpp]: :+1:

File: src/server/qgsserverprojectutils.cpp (hunk 2 above)
Comment (pathmapper) [reply src/server/qgsserverprojectutils.cpp]: :+1:

**Comment by elpaso** [comment]
LGTM thank you! There are still a few failing test that needs fixed and then we can merge.

A minor note: it's generally better to mark variables that do not change as `const`, can you please check your code for any missing `const` qualifier on variables?

**Comment by pathmapper** [comment]
Thank you for your guidance!

Now the Travis build is green but `OGC tests for QGIS Server` fails with:

`Error: The schema does not define xs:complexType arcType correctly.`

For me it looks like it's unrelated because the last two commits shouldn't cause this error and the OGC tests never failed for previous commits of this PR.

I'm going to close and re-open here to trigger the tests again.

**Review by elpaso — APPROVED** [review]
LGTM, thank you!

**Comment by elpaso** [comment]
CC @rldhont if you don't have any comment, I'd merge this one.

**Review by rldhont — APPROVED** [review]
Thanks

**Pull request merged**